src/test/java/tw/commonground/backend/service/fact/FactServiceTest.java

```diff
@@ -0,0 +1,189 @@
+        assertEquals(1, savedEntities.size());
+        assertEquals("https://www.github.com", savedEntities.getFirst().getUrl());
+        assertEquals("Github", savedEntities.getFirst().getTitle());
+        assertEquals("https://www.github.com/favicon.ico", savedEntities.getFirst().getFavicon());
+    }
+
+    @Test
+    void testGetUrlDetails_withDocumentWithoutIcon() throws Exception {
+        Document mockDocument = new Document("https://www.github.com");
+        mockDocument.title("Github");
+
+        FactService spyFactService = Mockito.spy(new FactService(factRepository, referenceRepository));
+        when(spyFactService.getDocument("https://www.github.com")).thenReturn(mockDocument);
+
+        ReferenceEntity referenceEntity = spyFactService.getUrlDetails("https://www.github.com");
+        assertThat(referenceEntity.getUrl()).isEqualTo("https://www.github.com");
+        assertThat(referenceEntity.getTitle()).isEqualTo("Github");
+        assertThat(referenceEntity.getFavicon()).isEqualTo("");
+    }
+
+    @Test
+    void testGetUrlDetails_withIconInLink() throws Exception {
+        Document mockDocument = new Document("https://www.github.com");
+        Element head = mockDocument.appendElement("head");
+        mockDocument.title("Github");
+        head.appendElement("link")
+                .attr("rel", "icon")
+                .attr("href", "https://www.github.com/favicon.ico");
+
+        FactService spyFactService = Mockito.spy(new FactService(factRepository, referenceRepository));
+        when(spyFactService.getDocument("https://www.github.com")).thenReturn(mockDocument);
+
+        ReferenceEntity referenceEntity = spyFactService.getUrlDetails("https://www.github.com");
+        assertThat(referenceEntity.getUrl()).isEqualTo("https://www.github.com");
+        assertThat(referenceEntity.getTitle()).isEqualTo("Github");
+        assertThat(referenceEntity.getFavicon()).isEqualTo("https://www.github.com/favicon.ico");
+    }
+
+    @Test
+    void testGetUrlDetails_withIconInMeta() throws Exception {
+        Document mockDocument = new Document("https://www.github.com");
+        Element head = mockDocument.appendElement("head");
+        mockDocument.title("Github");
+        head.appendElement("meta")
+                .attr("itemprop", "image")
+                .attr("content", "/favicon.ico");
+
+        FactService spyFactService = Mockito.spy(new FactService(factRepository, referenceRepository));
+        when(spyFactService.getDocument("https://www.github.com")).thenReturn(mockDocument);
+
+        ReferenceEntity referenceEntity = spyFactService.getUrlDetails("https://www.github.com");
+        assertThat(referenceEntity.getUrl()).isEqualTo("https://www.github.com");
+        assertThat(referenceEntity.getTitle()).isEqualTo("Github");
+        assertThat(referenceEntity.getFavicon()).isEqualTo("https://www.github.com/favicon.ico");
+    }
+
+    @Test
+    void testUrlHandling() {
```

Comment:
The test methods `testUrlHandling`, `testUrlHandling_withUrlWithoutProtocol`, `testUrlHandling_withEncodedUrl`, and `testUrlHandling_withEncodedUrlWithoutProtocol` differ only by one variable (input). Consider using `@Parameterized` to group these tests and reduce redundancy.

ref: https://www.baeldung.com/parameterized-tests-junit-5